src/System.Management.Automation/namespaces/FileSystemProvider.cs

```diff
@@ -1788,26 +1854,39 @@ protected override void InvokeDefaultAction(string path)
                                 return;
                             }
 
-                            // Once the recursion process has begun by being in this function,
-                            // we do not want to further recurse into directory symbolic links
-                            // so as to prevent the possibility of an endless symlink loop.
-                            // This is the behavior of both the Unix 'ls' command and the Windows
-                            // 'DIR' command.
-                            if (!InternalSymbolicLinkLinkCodeMethods.IsReparsePoint(recursiveDirectory))
+                            // We only want to recurse into symlinks if
+                            //  a) the user has asked to with the -FollowSymLinks switch parameter and
+                            //  b) the directory pointed to by the symlink has not already been visited,
+                            //     preventing symlink loops.
+                            if (tracker == null)
                             {
-                                bool hidden = false;
-                                if (!Force)
+                                if (InternalSymbolicLinkLinkCodeMethods.IsReparsePoint(recursiveDirectory))
                                 {
-                                    hidden = (recursiveDirectory.Attributes & FileAttributes.Hidden) != 0;
+                                    continue;
                                 }
-
-                                // if "Hidden" is explicitly specified anywhere in the attribute filter, then override
-                                // default hidden attribute filter.
-                                if (Force || !hidden || isFilterHiddenSpecified || isSwitchFilterHiddenSpecified)
+                            }
+                            else
+                            {
+                                if (tracker.IsPathVisited(recursiveDirectory.FullName))
                                 {
-                                    Dir(recursiveDirectory, recurse, depth - 1, nameOnly, returnContainers);
+                                    WriteWarning(StringUtil.Format(FileSystemProviderStrings.AlreadyListedDirectory,
+                                                                recursiveDirectory.FullName));
+                                    continue;
                                 }
```

Comment:
I don't agree with this. If we mark the directory has having been visited here, then if the test at line 1883 evaluates to false the `Dir` method will not be called and we will not visit the directory even though we have already claimed to have done so.